File: python/rpi-gpio-demo.py
#!/usr/bin/env python

from time import sleep
import gpio

def flags2str(flags):
    strs = []

    if (flags & gpio.PIN_INPUT):
        strs.append('INPUT')
    if (flags & gpio.PIN_OUTPUT):
        strs.append('OUTPUT')
    if (flags & gpio.PIN_OPENDRAIN):
        strs.append('OPENDRAIN')
    if (flags & gpio.PIN_PUSHPULL):
        strs.append('PUSHPULL')
    if (flags & gpio.PIN_TRISTATE):
        strs.append('TRISTATE')
    if (flags & gpio.PIN_PULLUP):
        strs.append('PULLUP')
    if (flags & gpio.PIN_PULLDOWN):
        strs.append('PULLDOWN')
    if (flags & gpio.PIN_INVIN):
        strs.append('INVIN')
    if (flags & gpio.PIN_INVOUT):
        strs.append('INVOUT')
    if (flags & gpio.PIN_PULSATE):
        strs.append('PULSATE')

    return ','.join(strs)

gpioc = gpio.controller()
max_pin = gpioc.max_pin

print "Max pin #: %d" % max_pin

# Dump information about all known pins
for p in range(0, max_pin + 1):
    pin = gpioc.pin(p)
    try:
        value = pin.value
    except gpio.GpioError as e:
        # pin with this number is unknown
        continue

    caps = flags2str(pin.caps)
    config = flags2str(pin.config)

    name = pin.name

    print "#%3d: %s, value=%d, config=<%s>, caps=<%s>" % (p, name, value, config, caps)

led_pin = gpioc.pin(16)

# pin LOW, LED is on
led_pin.value = gpio.LOW;
sleep(1);

# pin HIGH, LED is off
led_pin.value = gpio.HIGH;
sleep(1);

# blink LED several times
for i in range(100):
    led_pin.toggle()
    sleep(0.05)

File: perl/rpi-gpio-demo.pl
#!/usr/bin/perl -w
# Copyright (C) 2007 by Oleksandr Tymoshenko. All rights reserved.

use strict;
use Time::HiRes qw(usleep);
use GPIO;

my $controller = GPIO->new();
$controller->open() or die("Can't open GPIO controller dev: $!");
my $max_pin = $controller->get_max_pin();
print "Max pin #: $max_pin\n";

foreach my $p (0..$max_pin) {
    next if (!$controller->is_pin_valid($p));
    my $name = $controller->get_pin_name($p);
    my $value = $controller->get_pin_value($p);
    my $config = flags2str($controller->get_pin_config($p));
    my $caps = flags2str($controller->get_pin_caps($p));
    print "#$p: name=$name, value=$value, config=<$config>, caps=<$caps>\n";
}


# pin LOW, LED is on
$controller->set_pin_value(16, GPIO::LOW);
sleep(1);

# pin HIGH, LED is on
$controller->set_pin_value(16, GPIO::HIGH);
sleep(1);

# Blink LED
foreach (0..100) {
    $controller->toggle_pin_value(16);
    usleep(50000);
}

$controller->close();

# Helper functions
sub flags2str
{
    my $flags = shift;
    my @s = ();

    push @s, 'INPUT' if ($flags & GPIO::PIN_INPUT);
    push @s, 'OUTPUT' if ($flags & GPIO::PIN_OUTPUT);
    push @s, 'OPENDRAIN' if ($flags & GPIO::PIN_OPENDRAIN);
    push @s, 'PUSHPULL' if ($flags & GPIO::PIN_PUSHPULL);
    push @s, 'TRISTATE' if ($flags & GPIO::PIN_TRISTATE);
    push @s, 'PULLUP' if ($flags & GPIO::PIN_PULLUP);
    push @s, 'PULLDOWN' if ($flags & GPIO::PIN_PULLDOWN);
    push @s, 'INVIN' if ($flags & GPIO::PIN_INVIN);
    push @s, 'INVOUT' if ($flags & GPIO::PIN_INVOUT);
    push @s, 'PULSATE' if ($flags & GPIO::PIN_PULSATE);

    return join (',', @s);
}
